plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java

```diff
@@ -2254,6 +2270,46 @@ protected void configureVideoCard(VirtualMachineMO vmMo, VirtualMachineTO vmSpec
         }
     }
 
+    /**
+     * Sets GPU card details to the one provided in detail vgpu.type (if provided) on {@code vmConfigSpec}.
+     * @param vmMo
+     * @param vmSpec
+     * @param vmConfigSpec
+     * @throws RuntimeFaultFaultMsg
+     * @throws Exception
+     */
+    protected void configureGpuCard(final VirtualMachineMO vmMo, final VirtualMachineTO vmSpec, final VirtualMachineConfigSpec vmConfigSpec) throws Exception {
+        final String vGpuProfile = vmSpec.getGpuDevice().getVgpuType();
+
+        VirtualDevice pciPassthroughDevice = null;
+        HostMO hostMo = new HostMO(vmMo.getContext(), getHyperHost(vmMo.getContext()).getMor());
+
+        if (vGpuProfile.equalsIgnoreCase(GPU.GPUType.passthrough.toString())) {
+            // Fetch the available direct PCI graphics card available for allocation
+            String pciDeviceId = hostMo.getPciIdForAvailableDirectPciPassthroughDevice();
+            VirtualMachinePciPassthroughInfo hostPciDeviceInfo = hostMo.getHostPciDeviceInfo(pciDeviceId);
+
+            if (hostPciDeviceInfo == null) {
+                final String msg = "Unable to find free passthrough GPU to be allocated to VM. Please check the host hardware configuration.";
+                s_logger.error(msg);
+                throw new Exception(msg);
+            }
+
+            pciPassthroughDevice = hostMo.prepareDirectPciPassthroughDevice(hostPciDeviceInfo);
+        } else {
+            pciPassthroughDevice = hostMo.prepareSharedPciPassthroughDevice(vGpuProfile);
+        }
+
+        if (pciPassthroughDevice != null) {
```

Comment:
Is it possible for this `pciPassthroughDevice ` to be null after the previous IF/ELSE?

Reply:
@rafaelweingartner, I just cross-checked, It's not possible because we initialize the object in hostMo prepareDirectPciPassthroughDevice and prepareSharedPciPassthroughDevice If/Else statement. It's redundant but I feel it's safe to check. Do you want me to remove?  Thanks!!

Reply:
Well, I understand the idea to code avoiding/deflecting null pointer exceptions. However, cases like this when it is not possible to happen a null pointer exception. I would not do the check/conditional; otherwise, it feels that for every time we use an object we would have to add a conditional checking if that object is a null value or not.

I would not force (suggest strongly) you to remove this because it does not harm the code, but I called your attention to this matter. I know that it might seem nice that we are deflecting null pointers, but if they do not happen at all, we are not deflecting anything ;)

It is a personal choice (so the power is yours here).